sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuApproximatePercentile.scala

```diff
@@ -0,0 +1,171 @@
+
+  // the update expression will create a t-digest (List[Struct[Double, Double])
+  override lazy val updateExpressions: Seq[Expression] =
+    new CudfTDigest(inputBuf,
+      percentageExpression,
+      accuracyExpression) :: Nil
+
+  // the merge expression will merge t-digests
+  override lazy val mergeExpressions: Seq[GpuExpression] =
+    new CudfTDigest(outputBuf,
+      percentageExpression,
+      accuracyExpression) :: Nil
+
+  // the evaluate expression will compute percentiles based on a t-digest
+  override lazy val evaluateExpression: Expression = {
+    ApproxPercentileFromTDigestExpr(outputBuf, percentiles, child.dataType)
+  }
+
+  // inputBuf represents the initial aggregation buffer
+  protected final lazy val inputBuf: AttributeReference =
+    AttributeReference("inputBuf", CudfTDigest.dataType)()
+
+  // outputBuf represents the merged aggregation buffer
+  protected final lazy val outputBuf: AttributeReference =
+    inputBuf.copy("outputBuf", CudfTDigest.dataType)(inputBuf.exprId, inputBuf.qualifier)
+
+  override def nullable: Boolean = false
+
+  // Mark as lazy so that percentageExpression is not evaluated during tree transformation.
+  private lazy val (returnPercentileArray, percentiles) =
+    percentageExpression match {
+      case GpuLiteral(v, _) => makeArray(v)
+      case _ => makeArray(percentageExpression.eval())
+    }
+
+  private def makeArray(v: Any): (Boolean, Either[Double, Array[Double]])  = v match {
+    // Rule ImplicitTypeCasts can cast other numeric types to double
+    case null => (false, Right(Array()))
+    case num: Double => (false, Left(num))
+    case arrayData: ArrayData => (true, Right(arrayData.toDoubleArray()))
+  }
+
+  // The result type is the same as the input type.
+  private lazy val internalDataType: DataType = {
+    if (returnPercentileArray) ArrayType(child.dataType,
+      containsNull = false) else child.dataType
+  }
+
+  override def dataType: DataType = internalDataType
+
+  override def children: Seq[Expression] = Seq(child, percentageExpression, accuracyExpression)
+}
+
+case class ApproxPercentileFromTDigestExpr(
+    child: Expression,
+    percentiles: Either[Double, Array[Double]],
+    finalDataType: DataType)
+  extends GpuExpression {
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val expr = child.asInstanceOf[GpuExpression]
+    withResource(expr.columnarEval(batch).asInstanceOf[GpuColumnVector]) { cv =>
+
+      percentiles match {
+        case Left(p) =>
+          // For the scalar case, we still pass cuDF an array of percentiles
+          // (containing a single item) and then extract the child column from the resulting
+          // array and return that (after converting from Double to finalDataType
+          withResource(cv.getBase.approxPercentile(Array(p))) { percentiles =>
+            withResource(percentiles.getChildColumnView(0)) { childView =>
```

Comment:
What happens if there is a null at the list level? Is that guaranteed to never happen? If it did we could end up skipping that row in the child data column entirely.  I personally would rather see us call `extractListElement` which is guaranteed to work in all cases, even if this is technically safe to do.